File: backend/engine/events.py
"""Structured events emitted by the engine.

The engine never prints or logs. It returns events, and the CLI, the server and
the replay file all render the same stream. Every event is a plain dataclass
with `to_dict()` so it can go straight over the wire.
"""

from __future__ import annotations

from dataclasses import asdict, dataclass, field

from .constants import COLOR_NAMES


class Event:
    kind: str = "event"

    def to_dict(self) -> dict:
        d = asdict(self)  # type: ignore[call-overload]
        d["kind"] = self.kind
        if "color" in d and isinstance(d["color"], int) and d["color"] >= 0:
            d["color"] = COLOR_NAMES[d["color"]]
        return d


@dataclass
class Draft(Event):
    """A player took a color from a source and placed it."""

    player: int
    source: int
    color: int
    count: int
    dest: int
    placed: int  # tiles that landed on the staging row
    overflow: int  # tiles that landed on the penalty row
    to_discard: int  # tiles that fell off the end of a full penalty row
    took_first_token: bool
    kind: str = field(default="draft", init=False)


@dataclass
class TileScored(Event):
    """One tile moved onto the grid during settling."""

    player: int
    row: int
    col: int
    color: int
    points: int
    horizontal: int
    vertical: int
    kind: str = field(default="tile_scored", init=False)


@dataclass
class PenaltyApplied(Event):
    player: int
    tiles: int
    points: int  # negative or zero, before clamping at 0
    score_after: int
    kind: str = field(default="penalty", init=False)


@dataclass
class RoundEnd(Event):
    round_num: int
    scores: list[int]
    kind: str = field(default="round_end", init=False)


@dataclass
class RoundStart(Event):
    round_num: int
    first_player: int
    bag_refilled: bool
    short_displays: int  # displays that could not be filled to 4 (bag and discard empty)
    kind: str = field(default="round_start", init=False)


@dataclass
class BonusAwarded(Event):
    player: int
    rows: int
    columns: int
    colors: int
    points: int
    kind: str = field(default="bonus", init=False)


@dataclass
class GameEnd(Event):
    scores: list[int]
    winner: int | None
    draw: bool
    kind: str = field(default="game_end", init=False)

File: src/engine/events.ts
/**
 * Structured events emitted by the engine.
 *
 * The engine never logs. It returns events, and the UI, the replay file and the
 * parity vectors all render the same stream. Serialized shapes match
 * `backend/engine/events.py` exactly: `color` becomes a color name, and `kind`
 * is the discriminator.
 */

import { COLOR_WIRE_NAMES } from './constants';

export interface Draft {
  kind: 'draft';
  player: number;
  source: number;
  color: number;
  count: number;
  dest: number;
  /** tiles that landed on the staging row */
  placed: number;
  /** tiles that landed on the penalty row */
  overflow: number;
  /** tiles that fell off the end of a full penalty row */
  to_discard: number;
  took_first_token: boolean;
}

export interface TileScored {
  kind: 'tile_scored';
  player: number;
  row: number;
  col: number;
  color: number;
  points: number;
  horizontal: number;
  vertical: number;
}

export interface PenaltyApplied {
  kind: 'penalty';
  player: number;
  tiles: number;
  /** negative or zero, before clamping the score at 0 */
  points: number;
  score_after: number;
}

export interface RoundEnd {
  kind: 'round_end';
  round_num: number;
  scores: number[];
}

export interface RoundStart {
  kind: 'round_start';
  round_num: number;
  first_player: number;
  bag_refilled: boolean;
  /** displays that could not be filled to 4 (bag and discard both empty) */
  short_displays: number;
}

export interface BonusAwarded {
  kind: 'bonus';
  player: number;
  rows: number;
  columns: number;
  colors: number;
  points: number;
}

export interface GameEnd {
  kind: 'game_end';
  scores: number[];
  winner: number | null;
  draw: boolean;
}

export type GameEvent =
  | Draft
  | TileScored
  | PenaltyApplied
  | RoundEnd
  | RoundStart
  | BonusAwarded
  | GameEnd;

/**
 * Wire form of an event. Mirrors the Python `Event.to_dict`: field order aside,
 * the only transformation is the int `color` becoming its name.
 */
export function eventToDict(event: GameEvent): Record<string, unknown> {
  const out: Record<string, unknown> = { ...event };
  if (typeof out.color === 'number' && (out.color as number) >= 0) {
    out.color = COLOR_WIRE_NAMES[out.color as number];
  }
  return out;
}
